### DISCUSSION TITLE
MinerX FF Edition: Curio Storage - Testing Assignment Week 2

### DISCUSSION THREAD

**Discussion by Reiers** [filecoin-project/lotus#12024] [Storage Provider]
![Curio_MinerX_Banner_Week2](https://github.com/filecoin-project/lotus/assets/63351350/4b2a1408-3fbd-456c-88eb-f784ea518c05)

### Hello MinerX Fellows !

Week 2 is focused on getting hands-on experience with Curio Storage.
Everyone should either deploy or migrate an existing miner on calibration. 
Follow the guide below. Looking forward to the feedback and answers. Thank you!

## Requirement 
- OS: Linux
- Yugabyte

Hardware | Specification
-- | --
CPU | 8-core processor
CPU Support | Models with support for Intel SHA Extensions (AMD since Zen microarchitecture or Intel since Ice Lake) will significantly speed up the processes.
RAM | 256 GiB RAM + Swap
GPU | Nvidia GPU with at least 11GB VRAM
Disk | 2 TB NVMe disk

:rocket: This post if you are participating in this testing round.
If you aren't able to join for some reason, please reach out to me @Reiers here or on Slack!

### 1. Install Curio

```
git clone https://github.com/filecoin-project/lotus.git curio-calibration
cd curio-calibration
git checkout master
make clean && RUSTFLAGS="-C target-cpu=native -g" FFI_BUILD_FROM_SOURCE=1 make calibnet
sudo make install
```

### 2. Install Boost

```
git clone https://github.com/filecoin-project/boost boost-calibration
cd boost-calibration
git checkout feat/curio
make clean build
sudo make install
```

### 3. Install Yugabyte (if you haven't already)

❗If you have already set up a YugabyteDB for Boost then you can reuse the same YugabyteDB instance for Curio. You can skip directly to [migrating from Lotus-Miner to Curio](https://lotus.filecoin.io/storage-providers/curio/setup/#migrating-from-lotus-miner-to-curio) or [Initializing new Curio Miner](https://lotus.filecoin.io/storage-providers/curio/setup/#initiating-a-new-curio-cluster).

- https://docs.yugabyte.com/

### 4. Start syncing Lotus

```
aria2c -x5 https://forest-archive.chainsafe.dev/latest/calibnet/
lotus daemon --import-snapshot <snapshot-file>
```

### 5. Create or use existing addresses

- Owner: `lotus wallet new`
- Worker: `lotus wallet new bls`

For tFIL and DC:
- https://faucet.calibnet.chainsafe-fil.io/

### 6. Migrating from `lotus-miner` to `curio`

- `curio guided-setup`
- `curio test window-post task`

### 7. (Optional) Deploy a new Curio Storage Miner

- `curio guided-setup`

Docs: https://lotus.filecoin.io/storage-providers/curio/setup/#migrating-from-lotus-miner-to-curio

**Reply with:**
- [ ] - 🦭 If you have successfully sealed some CC sectors.
- [ ] - 🤝 If you have successfully made some storage deals.
- [ ] - 📁 Made a f05/market Deal.
- [ ] - 📁 Made a FIL+ Market Deal.
- [ ] - 📁 Made a FIL+ DDO Deal.
- [ ] - 🗄️ Put everything above in one sector. 
- [ ] - :open_file_folder: If you have successfully had some retrieval deals.
- [ ] - :computer: If you passed your first WindowPoSt.

:exclamation: Create [an issue](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Fbug&template=bug_report.yml) if you think there is a bug, don't forget to attach the logs! 🐛 

If you have any questions or need help, join:
- For dev questions : [#fil-curio-dev](https://filecoinproject.slack.com/archives/C06GD1SS56Y) 
- For general help : [#fil-curio-help](https://filecoinproject.slack.com/archives/C06LF5YP8S3)
- Calibraion net help : [#fil-net-calibration-discuss](https://filecoinproject.slack.com/archives/C01D42NNLMS)

**Comment by johansealstorage** [comment]
@Reiers , are there any hints or advise on deploying and configuring curio on ESPA type hardware?  In other words, CPU servers with SHA (no gpu) and GPU servers with no SHA but with multiple GPU's.

**Reply by Reiers** [reply]
**Great question!** 

You shouldn't have to do any tweaking with flags or add environment to your sealing pipeline - add the server to the curio cluster, let's say you have a "worker", point it at the DB and run `curio run --layers <the ones you want>` 

### CPU Servers with SHA (no GPU):

Focus on tasks that don't require GPU, such as sealing and post tasks.
Example command: `curio run --layers=seal`.

### GPU Servers with Multiple GPUs (no SHA):

Use for GPU-intensive tasks like SDR trees and PoRep proof generation.
Example command: `curio run --layers=seal-gpu,seal-snark`.

https://lotus.filecoin.io/storage-providers/curio/config/#layer-stacking

You can control everything from the UI, things that used to be environment: 
https://lotus.filecoin.io/storage-providers/curio/config/#pre-built-layers

**Reply by johansealstorage** [reply]
Still getting the hang of layers 😁

If `EnableSealSDRTrees`  is true in the layer `seal` and in `seal-gpu`, would it be executed on CPU Server or GPU server - or any/both?  
I guess the question is which curio node will grab that task?

**Reply by magik6k** [reply]
If you enable that task on both, the machines will have a ~50/50 chance of grabbing the task (assuming one machine with each layer)

(Once we land https://github.com/filecoin-project/lotus/pull/11796 the machine with data locally will be much more likely to get the Trees task)

**Reply by magik6k** [reply]
Note that you likely don't want the SDRTrees task enabled on the CPU machines, that task is equal to PreCommit2 in lotus-miner (except we also do TreeD here instead of in PreCommit1), so it really benefits from GPUs.

**Comment by SBudo** [comment]
- [x]  🦭 If you have successfully sealed some CC sectors.
- [x]  🤝 If you have successfully made some storage deals.
- [x]  📁 Made a f05/market Deal.
- [x]  📁 Made a FIL+ Market Deal.
- [x]  📁 Made a FIL+ DDO Deal.
- [ ]  🗄️ Put everything above in one sector.
- [x]  📂 If you have successfully had some retrieval deals.
- [x]  💻 If you passed your first WindowPoSt.

**Reply by SBudo** [reply]
It would be good to add the following features:

- Abort sealing jobs
- Update a sector state (to send it back in earlier stage of sealing process)
- Remove sectors (before they are on the chain)
- Add a filter on the PoRep pipeline page to filter out sectors that are proving, as to only view the sectors that are in the process of being sealed or in error

**Comment by stuberman** [comment]
- [x]  - 🦭 If you have successfully sealed some CC sectors.
- [x]  - 🤝 If you have successfully made some storage deals.
- [x]  - 📁 Made a f05/market Deal.
- [x]  - 📁 Made a FIL+ Market Deal.
- [x]  - 📁 Made a FIL+ DDO Deal.
- [ ]  - 🗄️ Put everything above in one sector.
- [x]  - 📂 If you have successfully had some retrieval deals.
- [x]  - 💻 If you passed your first WindowPoSt.

**Reply by stuberman** [reply]
**EDIT: I was fixed this by using default paths and not trying to set a custom path for data directories.** 

Having trouble integrating Boost, keep getting this error:

I run: `boostd-data run yugabyte --hosts 192.168.1.96 --connect-string="postgresql://yugabyte:yugabyte@192.168.1.96:5433?sslmode=disable" --addr 0.0.0.0:8044`

I get a response: 

> 2024-06-01T23:15:45.783Z	INFO	boostd-data-yb	yugabyte/setup.go:22	creating cassandra keyspace idx
> 2024-06-01T23:15:46.162Z	INFO	boostd-data-yb	yugabyte/setup.go:38	creating cassandra tables
> Error: starting yugabyte store: starting local index directory service: creating cassandra tables: executing
> CREATE TABLE IF NOT EXISTS PayloadToPieces (
>     PayloadMultihash BLOB,
>     PieceCid BLOB,
>     -- PayloadMultihash is the partition key: the key that determines which
>     -- node the row is sent to.
>     -- PieceCid is the clustering key: the data is sorted by this key.
>     -- We want both of these to be in the primary key because
>     -- the piece cids should be a unique set.
>     PRIMARY KEY (PayloadMultihash, PieceCid)
> )
> code=2200; message=Invalid Table Definition. Error creating table idx.payloadtopieces on the master: Not enough tablet servers in the requested placements. Need at least 1, have 0
> CREATE TABLE IF NOT EXISTS PayloadToPieces (
>                            ^^^^^^^^^^^^^^^
>     PayloadMultihash BLOB,
>     PieceCid BLOB,
>     -- PayloadMultihash is the partition key: the key that determines which
>     -- node the row is sent to.
>     -- PieceCid is the clustering key: the data is sorted by this key.
>     -- We want both of these to be in the primary key because
>     -- the piece cids should be a unique set.
>     PRIMARY KEY (PayloadMultihash, PieceCid)
> )
>  (ql error -302)
> 

`git show`

> commit cd7c946232662ddc6ff8f5be13dc68ea743c7880 (HEAD -> feat/curio, origin/feat/curio)

**Reply by stuberman** [reply]
I am finally able to run Boost now. 
Now I need to get some deals, any suggestions?

**Reply by stuberman** [reply]
As of June 5, I see these significant problems with the code:

- WinningPosts fail
- Inability to store Yugabyte data outside of /home directory (even though Yugabyte has flags for data paths)
- Incomplete documentation and/or features (how to create control addresses in Curio as example)

**Reply by stuberman** [reply]
With latest commits able to seal deals but commit message fails on Calibnet when multiple deals are in a sector of type verified and unverified due to gas error. I am told this is not an issue on mainnet.

**Reply by stuberman** [reply]
Next, planning to do a reinstall from scratch with new GA build and test.

**Comment by herrehesse** [comment]
* [ ]   - 🦭 If you have successfully sealed some CC sectors.
* [ ]   - 🤝 If you have successfully made some storage deals.
* [ ]   - 📁 Made a f05/market Deal.
* [ ]   - 📁 Made a FIL+ Market Deal.
* [ ]   - 📁 Made a FIL+ DDO Deal.
* [ ]   - 🗄️ Put everything above in one sector.
* [ ]   - 📂 If you have successfully had some retrieval deals.
* [ ]   - 💻 If you passed your first WindowPoSt.

**Comment by NDLABS-Leo** [comment]
> * [x]   - 🦭 If you have successfully sealed some CC sectors.
> 
>     * [ ]   - 🤝 If you have successfully made some storage deals.
> 
>     * [ ]   - 📁 Made a f05/market Deal.
> 
>     * [ ]   - 📁 Made a FIL+ Market Deal.
> 
>     * [ ]   - 📁 Made a FIL+ DDO Deal.
> 
>     * [ ]   - 🗄️ Put everything above in one sector.
> 
>     * [ ]   - 📂 If you have successfully had some retrieval deals.
> 
>     * [x]   - 💻 If you passed your first WindowPoSt.

![图片](https://github.com/filecoin-project/lotus/assets/104019147/421dd6a3-53c0-45df-b8b0-c2821a6547af)

**Comment by lei7337** [comment]
- [x] - 🦭 If you have successfully sealed some CC sectors.
 - 🤝 If you have successfully made some storage deals.
 - 📁 Made a f05/market Deal.
 - 📁 Made a FIL+ Market Deal.
 - 📁 Made a FIL+ DDO Deal.
 - 🗄️ Put everything above in one sector.
 - 📂 If you have successfully had some retrieval deals.
  - [x] -  💻 If you passed your first WindowPoSt.
  Miner ID [t0120833](https://calibration.filfox.info/en/address/t0120833)
